### ISSUE TITLE
"Limit in subquery is currently unsupported" - but it's a reference table

### ISSUE THREAD

**Issue by seamusabshere** [citusdata/citus#2393]
## The problem

Here's my query:

```
UPDATE distributed_table
SET
  nearest_id = (
    SELECT reference_table.id
    FROM reference_table
    ORDER BY distributed_table.the_geom <-> reference_table.the_geom
    LIMIT 1
 )
 ```

I get this error, but I think it's wrong:

```
ERROR:  cannot push down this subquery
DETAIL:  Limit in subquery is currently unsupported when a subquery references a column from another query
```

:star: Since the entire reference table exists on every shard, the answer to the subquery will be the same on every shard.

## My solution (which works but I don't want it)

```
WITH
geoms AS (
  SELECT
    id,
    shard_key,
    the_geom
  FROM distributed_table
),
closest AS (
  SELECT
    geoms.id,
    geoms.shard_key,
    (
      SELECT reference_table.id
      FROM reference_table
      ORDER BY geoms.the_geom <-> reference_table.the_geom
      LIMIT 1
    ) AS nearest_id
  FROM geoms
)
UPDATE distributed_table
SET
  nearest_id = closest.nearest_id
FROM closest
WHERE
      closest.id = distributed_table.id
  AND closest.shard_key = distributed_table.shard_key
```

**Comment by marcocitus** [comment]
Which version of Citus are you using? This seems to work on recent versions.

**Comment by seamusabshere** [comment]
@marcocitus 7.5

**Comment by onderkalaci** [comment]
It'd be a nice improvement to allow `LIMIT` in subqueries if we're not selecting from any distributed tables.

What about re-writing like the following:
```SQL
WITH max_distance_id AS
(
  SELECT * FROM
  (
    SELECT 
      reference_table.id, distributed_table.id as dist_id, distributed_table.the_geom <-> reference_table.the_geom as max_dist
    FROM
      reference_table, distributed_table
    ORDER BY
      max_dist
  ) as foo

  LIMIT 1
)
UPDATE 
  distributed_table 
SET 
  nearest_id = max_distance_id.id 
FROM 
  max_distance_id 
WHERE 
  distributed_table.id = max_distance_id.dist_id;
```

**Comment by marcocitus** [comment]
Ah, I missed the join in the ORDER BY. That makes it a correlated subquery which is usually tricky in combination with the LIMIT. Subqueries with LIMIT typically cannot be pushed down to the shards and subqueries with joins to outer tables cannot be executed separately either. However, we could push down subqueries on reference tables that have LIMIT *and* ORDER BY (to make results deterministic), but currently don't.

**Comment by marcocitus** [comment]
fwiw, one way I ran into which avoids CTEs involves adding a dummy column to `reference_table` to avoid some planner idiosyncracies, but it does become a cubic operation.

```sql
ALTER TABLE reference_table ADD COLUMN zero int default 0;   
UPDATE distributed_table d SET nearest_id = ref.id
FROM reference_table ref WHERE ref.the_geom <-> d.the_geom = (
  SELECT min(r.the_geom <-> d.the_geom) AS distance FROM reference_table r GROUP BY zero
);
```

**Comment by onderkalaci** [comment]
> hey @marcocitus is this a bug? thank you!

Let me share my thoughts on this. I wouldn't call this as a bug. That'd be an improvement that Citus currently doesn't implement. 

Also, note that I'm skeptical about allowing these types of correlated subqueries. Correlated subqueries could be very complicated** and require some more detailed investigation.

** you could have correlations in many different parts of the query (plus the correlation could be originated from any level of outer queries), and we should ensure that they always produce correct results.

**Comment by seamusabshere** [comment]
@onderkalaci but if it's a reference table, and we have both ORDER and LIMIT, won't it always be the same on all?

**Comment by marcocitus** [comment]
> @onderkalaci but if it's a reference table, and we have both ORDER and LIMIT, won't it always be the same on all?

I think we agree on that. A correlated subquery can be seen as a function that gets called for every row in the outer table. As long as that function is immutable and does not require data from other nodes, which is true in the reference table + order by + limit + no mutable functions + no mutable sub-subqueries case, we should be able to push it down. 

One question is more what the check for that looks like and where in the code it sits. We're a little weary of adding a lot of special cases to the planner instead of more generalised correlated subquery support.

Apart from how to make the query work, the more important question is how to make the query fast enough to be useful. As written, the query will always be O(D\*R) where D is the number of rows in the distributed table and R the number of rows in the reference table. For a 100M row distributed table with a 1M row reference table, the update would probably take 2 weeks on a sizable Citus cluster (and several years on PostgreSQL). I'm wondering whether there is a version of this subquery that can use a GiST index to make it O(D\*log R) (like this [single point example](https://boundlessgeo.com/2011/09/indexed-nearest-neighbour-search-in-postgis/)) and what the check for that query would look like, since it would reduce the query time to a few seconds on Citus.

**Comment by seamusabshere** [comment]
> I'm wondering whether there is a version of this subquery that can use a GiST index

@marcocitus For what it's worth, PostGIS will use a gist index for the `<->`, otherwise this would be useless

**Comment by seamusabshere** [comment]
@marcocitus is this on the roadmap?

**Comment by marcocitus** [comment]
Fixed by #4385
